keyboards/planck/keymaps/juno/keymap.c

```diff
@@ -0,0 +1,458 @@
+	_______, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11,  KC_F12,
+	_______, _______, _______, XXXXXXX, MO(_DPADNUM),
+												 XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX
+)
+
+
+};
+
+#ifdef AUDIO_ENABLE
+  float plover_song[][2]     = SONG(PLOVER_SOUND);
+  float plover_gb_song[][2]  = SONG(PLOVER_GOODBYE_SOUND);
+  
+  // Borrowing audio from unused audio
+  // Caps Lock on and off sound too similar
+  
+  float caps_song_on[][2] = SONG(NUM_LOCK_ON_SOUND);
+  float caps_song_off[][2] = SONG(SCROLL_LOCK_ON_SOUND);
+
+  float dpad_song_on[][2] = SONG(ZELDA_PUZZLE);
+  float dpad_song_off[][2] = SONG(SONIC_RING);
+#endif
+
+uint32_t layer_state_set_user(uint32_t state) {
+
+  // LED control, lighting up when Fn layer is activated
+  switch (biton32(state)) {
+    case _QWERTY:
+      backlight_set(0);
+      break;
+    case _FN1:
+      backlight_set(3);
+      break;
+	case _DPAD:
+      backlight_set(3);
+      break;
+  }
+
+  return update_tri_layer_state(state, _LOWER, _RAISE, _ADJUST);
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  switch (keycode) {
+    case QWERTY:
+      if (record->event.pressed) {
+        print("mode just switched to qwerty and this is a huge string\n");
+        set_single_persistent_default_layer(_QWERTY);
+      }
+      return false;
+      break;
+    case COLEMAK:
+      if (record->event.pressed) {
+        set_single_persistent_default_layer(_COLEMAK);
+      }
+      return false;
+      break;
+    case DVORAK:
+      if (record->event.pressed) {
+        set_single_persistent_default_layer(_DVORAK);
+      }
+      return false;
+      break;
+    case BACKLIT:
+      if (record->event.pressed) {
+        register_code(KC_RSFT);
+        #ifdef BACKLIGHT_ENABLE
+          backlight_step();
+        #endif
+        #ifdef KEYBOARD_planck_rev5
+          PORTE &= ~(1<<6);
+        #endif
+      } else {
+        unregister_code(KC_RSFT);
+        #ifdef KEYBOARD_planck_rev5
+          PORTE |= (1<<6);
+        #endif
+      }
+      return false;
+      break;
+    case PLOVER:
+      if (record->event.pressed) {
+        #ifdef AUDIO_ENABLE
+          stop_all_notes();
+          PLAY_SONG(plover_song);
+        #endif
+        layer_off(_RAISE);
+        layer_off(_LOWER);
+        layer_off(_ADJUST);
+        layer_on(_PLOVER);
+        if (!eeconfig_is_enabled()) {
+            eeconfig_init();
+        }
+        keymap_config.raw = eeconfig_read_keymap();
+        keymap_config.nkro = 1;
+        eeconfig_update_keymap(keymap_config.raw);
+      }
+      return false;
+      break;
+    case EXT_PLV:
+      if (record->event.pressed) {
+        #ifdef AUDIO_ENABLE
+          PLAY_SONG(plover_gb_song);
+        #endif
+        layer_off(_PLOVER);
+      }
+      return false;
+      break;
+	  
+	  
+// Play audio upon switching Caps Lock and custom layers
+
+	case KC_CAPS:
```

Comment:
You could actually use `led_set_user` for this, rather than a keycode. But either works.

Reply:
I asked the question on reddit (more than a year) and this was the best I could pull off. Do you mind showing a snippet of how this could have been done better?

Reply:
Absolutely:
https://docs.qmk.fm/#/custom_quantum_functions?id=example-led_set_user-implementation

Reply:
Also, a LOT has changed in a year.

Reply:
I think I get what your meant, but considering that I also use LED and audio indicator for my custom layer with T-inverted arrow cluster, which are definitely not in the API, I think my code is optimal in this case, with all being nested in one single instance of case-switch.

Let me know if it makes sense.